### PR TITLE
EOL `Global-Mask-Classes`

### PR DESCRIPTION
**Pull request by basil**
Reverts commit a9fc20d5b1882f8e29b91dc67379badbac106df6. I discovered this unmitigated complexity while reading this code recently. I do not know if it was ever used, but it is completely unused today: I checked both Jenkins and CloudBees sources and nothing came up. I cannot imagine any good reason to use this "feature." Better to rip it out now while we still can before someone decides to start using it in the future and then we are obligated to support this unusual use case. The code comment mentions something about database plugins, but I checked them and none of them are using this feature, nor is anything else in the Jenkins ecosystem using it today.

As a bonus, ripping out this feature allows us to simplify the implementation of `MaskingClassLoader` to use immutable data structures, making it easier to reason about the concurrency implications of that class.

If accepted, I will also file subsequent PRs to purge any references to this "feature" from the HPI plugin and documentation.

### Testing done

Started up Jenkins with `java -jar jenkins.war` with the `jakarta-mail-api` plugin installed, which exercises the (non-global) `Mask-Classes` code path and therefore the changed constructor. No errors.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Review by NotMyFault — APPROVED** [review]

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
